store/tikv/lock_resolver.go

```diff
@@ -578,10 +585,145 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
 			status.commitTS = cmdResp.CommitVersion
 			lr.saveResolved(txnID, status)
 		}
+		if cmdResp.UseAsyncCommit {
+			status.secondaries = cmdResp.GetSecondaries()
+		}
+
 		return status, nil
 	}
 }
 
+// resolveLockAsync resolves l assuming it was locked using the async commit protocol. Precondition: status.secondaries is non-empty.
+func (lr *LockResolver) resolveLockAsync(bo *Backoffer, l *Lock, status TxnStatus, lite bool, cleanRegions map[RegionVerID]struct{}) error {
+	tikvLockResolverCountWithResolveAsync.Inc()
+
+	regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, status.secondaries, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	commitTs := l.MinCommitTS
+	commitTsMutex := sync.Mutex{}
+	locks := make([]*Lock, len(regions)+1)
+	locks[0] = l
+	missingLock := false
+	errChan := make(chan error, len(regions))
+	count := 0
+
+	for regionID, keys := range regions {
+		count++
+		index := count
+		curRegionID := regionID
+		curKeys := keys
+
+		go func() {
+			for {
+				checkReq := &kvrpcpb.CheckSecondaryLocksRequest{
+					Keys:         curKeys,
+					StartVersion: l.TxnID,
+				}
+				req := tikvrpc.NewRequest(tikvrpc.CmdCheckSecondaryLocks, checkReq)
+				tikvLockResolverCountWithQueryCheckSecondaryLocks.Inc()
+				resp, err := lr.store.SendReq(bo, req, curRegionID, readTimeoutShort)
+				if err != nil {
+					errChan <- errors.Trace(err)
+					return
+				}
+				regionErr, err := resp.GetRegionError()
+				if err != nil {
+					errChan <- errors.Trace(err)
+					return
+				}
+				if regionErr != nil {
+					err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+					if err != nil {
+						errChan <- errors.Trace(err)
+						return
+					}
+					continue
+				}
+				if resp.Resp == nil {
+					errChan <- errors.Trace(ErrBodyMissing)
+					return
+				}
+				checkResp := resp.Resp.(*kvrpcpb.CheckSecondaryLocksResponse)
+
+				// Check each lock.
+				for _, lockInfo := range checkResp.Locks {
+					if lockInfo == nil {
+						// A lock is missing - the transaction must either be rolled back or committed.
+						commitTsMutex.Lock()
+						if !missingLock {
+							commitTs = checkResp.CommitTs
+						}
```

Comment:
I recommend to add a check if the `resp.CommitTs` equals to `commitTs` when `missingLock` is already `true`. This may help us to find bugs